### PR TITLE
:wrench: fixes #392 bmatrix calculation 

### LINKED ISSUE #392: Bmatrix formulation
** Bmatrix formulation **
```
// Gradient shapefn of the cell
  // dN/dx = [J]^-1 * dN/dxi
  Eigen::MatrixXd grad_shapefn = grad_sf * jacobian.inverse();
```
Change above to below
```
// Gradient shapefn of the cell
// dN/dx = [J]^-1 * dN/dxi
Eigen::MatrixXd grad_shapefn = grad_sf * (jacobian.inverse()).transpose();
```

** Check Bmatrix centroid too **

### PR DESCRIPTION
**Pull request by xinyiqian97**
Fixes #392 and tests for quadrilateral and hexahedron cases

### CODE REVIEW

File: include/quadrilateral_element.tcc [hunk 1]
```diff
@@ -323,14 +323,12 @@ inline std::vector<Eigen::MatrixXd>
...
-  Eigen::MatrixXd grad_shapefn = grad_sf * jacobian.inverse();
+  Eigen::MatrixXd grad_shapefn = grad_sf * (jacobian.inverse()).transpose();
 
   for (unsigned i = 0; i < Tnfunctions; ++i) {
     Eigen::Matrix<double, 3, Tdim> bi;
     // clang-format off
-    bi(0, 0) = grad_shapefn(i, 0); bi(0, 1) = 0.;
-    bi(1, 0) = 0.;                 bi(1, 1) = grad_shapefn(i, 1);
-    bi(2, 0) = grad_shapefn(i, 1); bi(2, 1) = grad_shapefn(i, 0);
+    Eigen::MatrixXd grad_shapefn = jacobian.inverse() * grad_sf.transpose();
```
Comment (ezrayst) [inline include/quadrilateral_element.tcc]: Do you mean to keep `bi` here instead of just `grad_shapefn`?

File: include/quadrilateral_element.tcc [hunk 2]
```diff
@@ -323,14 +323,12 @@ inline std::vector<Eigen::MatrixXd>
 
   // Gradient shapefn of the cell
   // dN/dx = [J]^-1 * dN/dxi
-  Eigen::MatrixXd grad_shapefn = grad_sf * jacobian.inverse();
+  Eigen::MatrixXd grad_shapefn = grad_sf * (jacobian.inverse()).transpose();
 
   for (unsigned i = 0; i < Tnfunctions; ++i) {
     Eigen::Matrix<double, 3, Tdim> bi;
     // clang-format off
```
Comment (ezrayst) [inline include/quadrilateral_element.tcc]: Can remove `// clang-format off` and on down below if you are doing this

**Review by ezrayst — COMMENTED** [review]
Fix #392

File: include/quadrilateral_element.tcc [hunk 3]
```diff
@@ -323,14 +323,12 @@ inline std::vector<Eigen::MatrixXd>
 
   // Gradient shapefn of the cell
   // dN/dx = [J]^-1 * dN/dxi
-  Eigen::MatrixXd grad_shapefn = grad_sf * jacobian.inverse();
+  Eigen::MatrixXd grad_shapefn = grad_sf * (jacobian.inverse()).transpose();
```
Comment (srhgk2) [inline include/quadrilateral_element.tcc:326]: Can you change this to`Eigen::MatrixXd grad_shapefn = jacobian.inverse() * grad_sf.transpose();`

File: include/quadrilateral_element.tcc (hunk 1 above)
Comment (srhgk2) [inline include/quadrilateral_element.tcc]: Remove this line.

File: include/quadrilateral_element.tcc [hunk 4]
```diff
@@ -323,14 +323,12 @@ inline std::vector<Eigen::MatrixXd>
...
 
   for (unsigned i = 0; i < Tnfunctions; ++i) {
     Eigen::Matrix<double, 3, Tdim> bi;
     // clang-format off
-    bi(0, 0) = grad_shapefn(i, 0); bi(0, 1) = 0.;
```
Comment (srhgk2) [inline include/quadrilateral_element.tcc]: `bi(0, 0) = grad_shapefn(0,i);`

File: include/quadrilateral_element.tcc [hunk 5]
```diff
@@ -323,14 +323,12 @@ inline std::vector<Eigen::MatrixXd>
...
-    bi(1, 0) = 0.;                 bi(1, 1) = grad_shapefn(i, 1);
```
Comment (srhgk2) [inline include/quadrilateral_element.tcc]: `bi(1, 1) = grad_shapefn(1,i);`

File: include/quadrilateral_element.tcc [hunk 6]
```diff
@@ -323,14 +323,12 @@ inline std::vector<Eigen::MatrixXd>
...
-    bi(2, 0) = grad_shapefn(i, 1); bi(2, 1) = grad_shapefn(i, 0);
```
Comment (srhgk2) [inline include/quadrilateral_element.tcc]: `bi(2, 0) = grad_shapefn(1, i); bi(2, 1) = grad_shapefn(0, i);`

**Comment by cw646** [comment]
Will we also need this fix for GIMP elements?

**Review by srhgk2 — APPROVED** [review]

**Pull request merged**